scripts/sonic-kdump-config

```diff
@@ -0,0 +1,632 @@
+    if curr_crash_kernel_mem == None:
+        lines[next_img_index] += " crashkernel=%s" % memory
+        changed = True
+        if verbose:
+            print("Added to grub.cfg: [ crashkernel=%s ]" % memory)
+    else:
+        if curr_crash_kernel_mem == memory:
+            if curr_crash_kernel_mem == crash_kernel_in_cmdline:
+                print("kdump is already enabled")
+            else:
+                changed = True
+        else:
+            lines[next_img_index] = lines[next_img_index].replace(curr_crash_kernel_mem, memory)
+            changed = True
+            if verbose:
+                print("Replace [%s] with [%s] in grub.cfg" % (curr_crash_kernel_mem, memory))
+
+    if changed:
+        rewrite_grub_cfg(lines, grub_cfg)
+
+    write_use_kdump(1)
+
+    return changed
+
+## Command: Enable kdump on Next image only
+#
+#  @param verbose If True, the function will display a few additional information
+#  @return        True is the grub configuration has changed, and False if it has not
+def cmd_kdump_config_next(verbose):
+
+    kdump_enabled = get_kdump_administrative_mode()
+    memory = get_kdump_memory()
+    num_dumps = get_kdump_num_dumps()
+    if verbose:
+        print("configDB: kdump_enabled=%d memory=[%s] num_nums=%d" % (kdump_enabled, memory, num_dumps))
+
+    if os.path.exists(grub_cfg):
+        return kdump_config_next_grub(verbose, kdump_enabled, memory, num_dumps)
+    else:
+        return False
+
+## Command: Disable kdump - Grub mode
+#
+#  @param verbose If True, the function will display a few additional information
+def kdump_disable_grub(verbose, kdump_enabled, memory, num_dumps):
+    write_use_kdump(0)
+
+    current_img = get_current_image();
+    if verbose:
+        print("Current image=[%s]\n" % current_img)
+    lines = [line.rstrip('\n') for line in open(grub_cfg)]
+    current_img_index = locate_image(lines, "loop=image-"+current_img)
+
+    changed = False
+    curr_crash_kernel_mem = search_for_crash_kernel(lines[current_img_index])
+    if curr_crash_kernel_mem == None:
+        print("kdump is already disabled")
+    else:
+        lines[current_img_index] = lines[current_img_index].replace("crashkernel="+curr_crash_kernel_mem, "")
+        changed = True
+        if verbose:
+            print("Removed [%s] in grub.cfg" % ("crashkernel="+curr_crash_kernel_mem))
+
+    if changed:
+        rewrite_grub_cfg(lines, grub_cfg)
+
+
+## Command: Disable kdump
+#
+#  @param verbose If True, the function will display a few additional information
+def cmd_kdump_disable(verbose):
+
+    kdump_enabled = get_kdump_administrative_mode()
+    memory = get_kdump_memory()
+    num_dumps = get_kdump_num_dumps()
+    if verbose:
+        print("configDB: kdump_enabled=%d memory=[%s] num_nums=%d" % (kdump_enabled, memory, num_dumps))
+
+    if os.path.exists(grub_cfg):
+        return kdump_disable_grub(verbose, kdump_enabled, memory, num_dumps)
```

Comment:
It's better than the previous version but I still feel some abstraction is lacking.
Basically with this design for every operation each bootloader has to be added inside a if/elif/elif. This is redundant and unnecessary.
Both the `kdump_disable_grub` and `kdump_config_next_grub` functions have a mix of common logic tied with grub specific code. Adding a new bootloader to this file would likely end up with a lot of copy pasting.

Could we instead just rely on a `get_bootloader()` function? This function would hold the if/elif/elif that returns the Bootloader instance or raise UnknownBootloaderError.
Each bootloader would then override the method of the base class Bootloader to add support.

Adding a new Bootloader would now just be about adding a new class definition and changing 2 lines in ` `get_bootloader` which is a much better design.

Reply:
@olivier-singla , can you address them?

Reply:
I am not sure more abstraction will make things easier or simpler when adding support for another boot loader. There is little code duplicated between `kdump_enable_grub`, `kdump_config_next_grub` and `cmd_kdump_disable_grub`. grub and u-boot (for instance) use a different mechanism: grub use a flat text file, where u-boot use environment variables.